### PR TITLE
petsc: when hdf5 depends on zlib - petsc+hdf5 should also be configur…

### PR DESCRIPTION
**Pull request by balay**
…ed with zlib

This issue came up with static builds

@keitat

### CODE REVIEW

File: var/spack/repos/builtin/packages/petsc/package.py [hunk 1]
```diff
@@ -114,6 +114,7 @@ class Petsc(Package):
     depends_on('metis@5:+int64', when='+metis+int64~double')
 
     depends_on('hdf5+mpi', when='+hdf5+mpi')
+    depends_on('zlib', when='+hdf5+mpi+zlib')
```
Comment (davydden) [inline var/spack/repos/builtin/packages/petsc/package.py]: there is no `zlib` variant in `petsc` so I don't see how `+zlib` can be triggered in `when`.

probably `depends_on('zlib', when='^hdf5+mpi')` is what you need.

**Comment by balay** [comment]
Hm - for some reason this did result in petsc picking up zlib in my
testing. [Perhaps some other dep caused it?]

My intention was to implment:

`if hdf5 is built with zlib - then petsc should also check for zlib
[to satisfy hdf5's dependency on zlib - otherwise there are unresolved
symbol errors when linking with hdf5].`

So is the following ok?

`depends_on('zlib', when='^hdf5+mpi+zlib')`

**Comment by davydden** [comment]
@balay 

> if hdf5 is built with zlib - then petsc should also check for zlib `depends_on('zlib@1.1.2:')`

but in Spack `hdf5` has unconditional dependency on `zlib`. So this should be enough
```
depends_on('zlib', when='+hdf5')
```

**Comment by balay** [comment]
On Wed, 8 Nov 2017, Denis Davydov wrote:

> if hdf5 is built with zlib - then petsc should also check for zlib `depends_on('zlib@1.1.2:')`

Ideally I should just query hdf5 to get is dependencies and use it [instead of also making petsc
depend on hdf5's dependencies.] But I won't worry about that now.

> but in Spack `hdf5` has unconditional dependency on `zlib`.

Yeah - I saw that - but didn't want to rely on this info in
petsc/package.py [what if it changes at some point - such code
dependencies across files usually cause bugs later.]. Hence I wanted
to add syntax matching:

"if hdf5 is built with zlib - then enable zlib in petsc configure"

> So this should be enough
> ```
> depends_on('zlib', when='+hdf5')
> ```

But not a big deal. Will push this change.

**Comment by davydden** [comment]
> Ideally I should just query hdf5 to get is dependencies and use it

I am quite certain you can, but I just don't know how. @tgamblin ping?

**Comment by balay** [comment]
Pushed a potential fix for https://github.com/spack/spack/pull/6189

**Comment by tgamblin** [comment]
@balay:

> Ideally I should just query hdf5 to get is dependencies and use it [instead of also making petsc
depend on hdf5's dependencies.] But I won't worry about that now.

Try:

```python
# This checks whether `zlib` is present in PETSc's DAG at all
if 'zlib' in spec:
    # do stuff for when `zlib` is present, e.g. do something with spec['zlib']
```

```python
# this checks specifically whether hdf5 depends on zlib, not if *anything* in 
# the DAG depends on zlib
# could also do spec['hdf5'].satisfies('^zlib') but I find it less readable in Python.
if 'zlib' in spec['hdf5']:
    # do stuff with spec['hdf5']; now we know it depends on zlib
```

**Comment by balay** [comment]
On Wed, 8 Nov 2017, Todd Gamblin wrote:

> @balay:
> 
> > Ideally I should just query hdf5 to get is dependencies and use it [instead of also making petsc
> depend on hdf5's dependencies.] But I won't worry about that now.
> 
> Try:
> 
> ```python
> # This checks whether `zlib` is present in PETSc's DAG at all
> if 'zlib' in spec:
>     # do stuff for when `zlib` is present, e.g. do something with spec['zlib']
> ```
> 
> ```python
> # this checks specifically whether hdf5 depends on zlib, not if *anything* in 
> # the DAG depends on zlib
> # could also do spec['hdf5'].satisfies('^zlib') but I find it less readable in Python.
> if 'zlib' in spec['hdf5']:
>     # do stuff with spec['hdf5']; now we know it depends on zlib
> ```

Thanks! I'll leave the code as is for now.

My though here is - PETSc shouldn't need to know or care about hdf5's
dependencies [if PETSc doesn't have an optional dependency on these
packages]. But it need that info just so that compiles don't break.

hdf5 can depend on zlib, szip [or bunch of other things]. And this
can go on recursively. [zlib can have its own dependencies etc..]

So checking for zlib explicitly in petsc/pacakge.py is not a good
general strategy.

[yeah petsc configure has interface for zlib - just to make it easy
for users to configure build hdf5 - and install petsc with hdf5]

so if petsc depends on [pkg-a, pkg-b, pkg-c] and pkg-a depends on
[pkg-aa,pkg-bb] pkg-b depends on [pkg-aa,pkg-cc], and pkg-c depend on
[pkg-cc] ultimately there should be  mechanism to say:

deps = [pkg-a, pkg-b, pkg-c]
pkgs = deps.pkgs() -> pkg-aa,pkg-bb,pkg-cc
CFLAGS=deps.inc()  -> -Ipkg-aa-inc -Ipkg-bb-inc -Ipkg-cc-inc
LDFLAGS=deps.lib() -> -lpkg-aa -lpkg-bb -lpkgcc

Now I should be able to do:

for pkg with petsc interface:
  if pkg in pkgs:  use --with-$pkg=1

And then I should be able to configure PETSc with:

./configure --with-pkg-a=1 --with-pkg-b=1 -with-pkg-cc=1 CFLAGS=$CFLAGS LDFLAGS=$LDFLAGS

Sure - every package need to support this way of configuring/building
(and I'm sure some don't) - and spack would need to know the inc and
lib values for each of the package it installs. Perhaps spack already
has some mechanisms to support this.


Note: the 2 other problems we've discussed fall into this problem space.
- -lm issue with plasma [it was due to dependency from some blas impls]
- -lgfortran issue @keitat currently has with plasma with static builds [again a dependency of blas]

**Pull request merged**